### ISSUE TITLE
thread-safe version of PlanningScene::decoupleParent?

### ISSUE THREAD

**Issue by simonschmeisser** [ros-planning/moveit#432]
### Description

We just noticed (due to crashes) that PlanningScene::decoupleParent is not thread-safe as it updates the RobotState of parent before cloning.

https://github.com/ros-planning/moveit/blob/4bc43c56ca2a5293f3ef8d55796ec586cfde5f69/moveit_core/planning_scene/src/planning_scene.cpp#L1111

```
if (!ftf_)
  {
    ftf_.reset(new SceneTransforms(this));
    ftf_->setAllTransforms(parent_->getTransforms().getAllTransforms());
}
```

https://github.com/ros-planning/moveit/blob/4bc43c56ca2a5293f3ef8d55796ec586cfde5f69/moveit_core/planning_scene/src/planning_scene.cpp#L688

```
const robot_state::Transforms& planning_scene::PlanningScene::getTransforms()
{
  getCurrentStateNonConst().update();
  return const_cast<const PlanningScene*>(this)->getTransforms();
}
```
(btw, getCurrentStateNonConst() calls update already)

We are testing many picks in parallel and need a copy of the current PlanningScene for each try so that we can remove the object of interest and attach it to the gripper. Of course we can serialize this cloning here, but it would be really nice to have a const/thread-safe version of decouple parent.

Also having three versions of getTransforms is a bit confusing, there is

* getTransforms const
* getTransforms (which is "semi-const" as it does the updating mentioned above)
* getTransformsNonConst

### Your environment
* ROS Distro: Kinetic
* OS Version: Ubuntu 16.04
* Source https://github.com/ros-planning/moveit/commit/4bc43c56ca2a5293f3ef8d55796ec586cfde5f69

**Comment by v4hn** [comment]
Sounds fully reasonable. The current situation is confusing indeed.

Could you provide a concrete patch for this against kinetic?
